Fixed #26421 -- Refactored ModelSignal to use Apps.lazy_model_operation()

AlexHill opened the pull request:
Currently `ModelSignal` has its own mechanism for resolving model strings for no good reason. This patch refactors it to use the mechanism in `Apps` instead.

There are two commits here.

The first is a slight refactor of `Apps.lazy_model_operation` and related code that
* expresses its workings in a clearer way
* improves the system's generality by allowing 0 or more models rather than 1 or more
* introduces a new check to detect unresolved reference and output nicer error messages, and removes similar code in `StateApps`

The second commit refactors `ModelSignal` to use `Apps.lazy_model_operation`.

charettes commented on django/apps/registry.py:
```diff
@@ -372,29 +372,37 @@ def lazy_model_operation(self, function, *model_keys):
-            self._pending_operations[model_key].append(function)
+        # Base case: no arguments, just execute the function.
+        if not model_keys:
+            function()
```
What about renaming `function` to `callback` in this method just like you did in `do_pending_operations`?

AlexHill replied:
I didn't actually mean for that change to make it – reverted for consistency. Unless there's a good reason not to use "function"?

charettes replied:
No particular reason, my argument was about consistency.

charettes commented on django/contrib/contenttypes/fields.py:
```diff
@@ -414,20 +414,23 @@ def contribute_to_class(self, cls, name, **kwargs):
+
+            def make_generic_foreign_order_accessors(related_model, model):
+                if corresponding_generics(self, model._meta.order_with_respect_to):
+                    make_foreign_order_accessors(model, related_model)
```
While you are around I think we should inline `corresponding_generics` to increase readability,

AlexHill replied:
I found the identical code further up in the definition of `GenericRelation`, and factored it out into a `_is_matching_generic_foreign_key()` method.

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -64,8 +64,10 @@ def extract_operation(obj):
-            args.extend(getattr(operation, 'args', []))
-            keywords.update(getattr(operation, 'keywords', {}))
+            # The or clauses here are redundant but work around a
+            # bug in functools.partial in certain Python versions.
```
I would say "certain Python 2.7 versions" so we know that it's fine to remove when dropping support for it.

AlexHill replied:
[The same bug was fixed in 3.5](https://mail.python.org/pipermail/python-checkins/2016-February/141214.html), so there may be circumstances where this can fail in Python 3 as well. Better leave it?

timgraham replied:
2.7/3.5?

AlexHill replied:
Yep wasn't thinking long-term enough. I've specified the exact versions and put the Python issue number in the comment.

charettes commented on django/contrib/contenttypes/fields.py:
```diff
@@ -328,14 +328,18 @@ def check(self, **kwargs):
+            field.ct_field == self.content_type_field_name and
+            field.fk_field == self.object_id_field_name
+        )
+
```
Great!

AlexHill commented:
All suggestions addressed – thanks for reviewing!

charettes commented on django/core/checks/model_checks.py:
```diff
@@ -119,6 +88,29 @@ def field_error(model_key, func, args, keywords):
+        params = {
+            'model': '.'.join(model_key),
+            'receiver': description,
+            'signal': signal_name
```
Add a trailing comma.

AlexHill replied:
Fixed in latest commit.

aaugustin commented on django/db/migrations/state.py:
```diff
@@ -240,44 +240,11 @@ def __init__(self, real_apps, models, ignore_swappable=False):
-        msg = "Unhandled pending operations for models:"
-        return "\n  ".join([msg] + [model_text(*i) for i in models_fields])
+        from django.core.checks.model_checks import _check_lazy_references
+        ignore = {make_model_tuple(settings.AUTH_USER_MODEL)} if ignore_swappable else set()
```
I think we should ignore _all_ swappable models, but that's another topic.

aaugustin commented on django/core/checks/model_checks.py:
```diff
@@ -63,3 +64,91 @@ def check_model_signals(app_configs=None, **kwargs):
+    def field_error(model_key, func, args, keywords):
+        error_msg = (
+            "The field %(field)s was declared with a lazy reference "
+            "to '%(model)s', which is not installed."
```
Unlike an "app", I wouldn't say that a model is "installed". Two solutions:

- check if the app is installed and adjust the error message accordingly: "app X isn't installed" or "app X doesn't provide model Y"
- just rephrase do "which doesn't exist"

AlexHill replied:
Ok, I've updated the PR to do the former.

aaugustin commented:
I don't know that code well enough to guarantee that the refactoring is correct, but generally speaking, this change seems sensible to me.

charettes commented on django/core/checks/model_checks.py:
```diff
@@ -31,35 +31,131 @@ def check_all_models(app_configs=None, **kwargs):
+
+@register(Tags.models)
+def check_lazy_references(app_configs=None, **kwargs):
+    return _check_lazy_references(apps)
```
It would have been more appropriate if `run_checks()` took an `Apps` instance as parameters instead of a list of `AppConfig`s.

AlexHill replied:
Yep. Especially since almost everywhere that argument is used in Django, it looks like this:

```python
checks.run_checks(app_configs=apps.get_app_configs())
```

charettes replied:
I guess we could refactor this in another commit/PR. From a quick look it seems possible to remove the `app_configs` argument from `run_checks()`, accept an `apps` argument instead and pass `apps=apps, app_configs=apps.get_app_configs()` to checks as we documented to always accept `**kwargs` just like signal receivers. The only part that might require deprecation is the `BaseCommand.check()` method.

AlexHill replied:
I think leaving both arguments could be a bit confusing – might it be better to deprecate `app_configs` in favour of calling `Apps.set_available_apps()` then passing an app registry? That way checks can use `apps.get_app_configs()` if they need to, and we enforce the constraint that the apps passed into checks are a subset of `INSTALLED_APPS`.

AlexHill commented:
Alright, should be good to go. I've just squashed this into the two original logical commits: the first making some accommodations for the `ModelSignal` refactor and the second implementing it.

AlexHill commented:
The docs test failure looks like an error in the test run:

> WARNING: intersphinx inventory u'http://sphinx-doc.org/objects.inv' not fetchable due to <class 'socket.error'>: [Errno 104] Connection reset by peer

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -63,3 +64,101 @@ def check_model_signals(app_configs=None, **kwargs):
+            'field': keywords['field'],
+            'model_error': app_model_error(model_key),
+        }
+        return Error(error_msg % params, obj=keywords['field'], hint=None, id='fields.E020')
```
Should be documented in `docs/ref/checks.txt`.

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -63,3 +64,101 @@ def check_model_signals(app_configs=None, **kwargs):
+            'model': '.'.join(model_key),
+            'model_error': app_model_error(model_key),
+        }
+        return Error(error_msg % params, obj=func, hint=None)
```
missing id?

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -63,3 +64,101 @@ def check_model_signals(app_configs=None, **kwargs):
+
+
+@register(Tags.models)
+def check_lazy_references(app_configs=None, **kwargs):
```
Is this a new check? Should it have some tests?

timgraham commented on django/db/models/utils.py:
```diff
@@ -7,12 +7,17 @@ def make_model_tuple(model):
+            model_tuple = app_label, model_name.lower()
+        else:
+            model_tuple = model._meta.app_label, model._meta.model_name
+        assert len(model_tuple) == 2
```
is it a concern that the assert won't run if running with `python -O`?

aaugustin replied:
It doesn't concern me — it just means that things could blow up in a more confusing way if the assert was silently skipped.

timgraham commented on django/db/models/utils.py:
```diff
@@ -7,12 +7,17 @@ def make_model_tuple(model):
+        return model_tuple
+    except (ValueError, AssertionError):
+        raise ValueError(
+            "Invalid model reference '%s'. String model references "
```
Is this change in the correct commit? The error message isn't updated in a test until the second commit.

AlexHill replied:
Yep: after the first commit, the error in the test still originates in [django/db/models/signals.py](https://github.com/django/django/pull/6359/files#diff-b9dad9123b06552b17edf4c85a802a5dL39). After the refactor in the second commit, the origin of the error is this function, so the message is updated.

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -83,6 +48,10 @@ def _check_lazy_references(app_registry, ignore=None):
         return []
 
+    from django.db.models import signals
+    model_signals = {signal: name for name, signal in vars(signals).items()
```
prefer hanging indent: 
```
model_signals = {
    signal: name for name, signal in vars(signals).items()
    if isinstance(signal, signals.ModelSignal)
}
```

AlexHill replied:
Done in 1d8f775

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -128,6 +97,29 @@ def field_error(model_key, func, args, keywords):
+            'signal': signal_name,
+            'model_error': app_model_error(model_key),
+        }
+        return Error(error_msg % params, obj=receiver.__module__, hint=None, id='signals.E001')
```
no need for `hint=None`

AlexHill replied:
Done in 6fcc9c5 and 2b93053

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -128,6 +97,29 @@ def field_error(model_key, func, args, keywords):
+    def signal_connect_error(model_key, func, args, keywords):
+        error_msg = (
+            "%(receiver)s was connected to the '%(signal)s' signal with a "
+            "lazy reference to the sender '%(model)s', but %(model_error)s."
```
looks like the wording of this message should be updated in `docs/ref/checks.txt`

AlexHill replied:
Updated in 1924d30

timgraham commented on django/core/checks/model_checks.py:
```diff
@@ -128,6 +97,29 @@ def field_error(model_key, func, args, keywords):
+        if isinstance(receiver, types.FunctionType):
+            description = "The '%s' function" % receiver.__name__
+        elif isinstance(receiver, types.MethodType):
+            description = "Bound method %s.%s" % (receiver.__self__.__class__.__name__, receiver.__name__)
```
add single quotes around  "%s.%s for consistency with other descriptions?

AlexHill replied:
Done in fa79a79

timgraham commented:
Tests aren't running and there are some flake8 errors after the first commit.
Some cosmetic edits for that commit:  http://dpaste.com/157M0TX

charettes commented on django/core/checks/model_checks.py:
```diff
@@ -31,35 +31,127 @@ def check_all_models(app_configs=None, **kwargs):
 
-@register(Tags.models, Tags.signals)
-def check_model_signals(app_configs=None, **kwargs):
+def _check_lazy_references(app_registry, ignore=None):
```
I think the norm is to use `apps` instead of `app_registry` to designate `Apps()` instances around the codebase.

AlexHill replied:
Done in 2223456 and 9b2a8a3

charettes commented on django/db/models/signals.py:
```diff
@@ -11,44 +12,15 @@ class ModelSignal(Signal):
     Signal subclass that allows the sender to be lazily specified as a string
     of the `app_label.ModelName` form.
     """
+    def connect(self, receiver, sender=None, weak=True, dispatch_uid=None, app_registry=None):
```
Ditto about the `app_registry` comment.

timgraham commented:
@AlexHill, do you think you'll have a chance to update this before the 1.10 alpha on May 16?

AlexHill commented:
Yep, I'll find some time this week.

Alex

On Sunday, 8 May 2016, Tim Graham <notifications@github.com> wrote:

> @AlexHill <https://github.com/AlexHill>, do you think you'll have a
> chance to update this before the 1.10 alpha on May 16?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/6359#issuecomment-217669078>
>

AlexHill commented:
Ok, I've updated this PR with the latest feedback. I've left all partial commits intact to keep the individual changes visible; I'll squash back down to two after review.

The two eventual commits will correspond to 5745c2d and 25e36d2. Both those pass isort, flake8, and the full test suite on my machine.

There remain questions about how to handle checks.

* There is some overlap between `fields.E300` (which checks for lingering strings and abstract models in `RelatedField`s) and the new `fields.E350`. Should we:
 * Not worry about it, have a misconfiguration raise both errors
 * Remove the string checking from E300
 * Something else?
* Several of the new checks return a different error message depending on whether or not an app referred to in a model reference is installed, as per [Aymeric's suggestion](https://github.com/django/django/pull/6359#discussion_r57953380). Should the two variants be two different codes in each case? For now, they are not – I've just included the text of both in the documentation.
* Should all the new checks just be in the `models` namespace, instead of in `signals`, `fields` etc? I think perhaps they should.

AlexHill commented:
[Test failures again seem unrelated to this code.](http://djangoci.com/job/pull-requests-trusty/6855/database=postgres,label=trusty-pr,python=python3.5/testReport/junit/auth_tests.test_hashers/TestUtilsHashPassArgon2/test_argon2_upgrade/) I'll rebase and see if that makes any difference.

timgraham commented on tests/model_validation/tests.py:
```diff
@@ -1,20 +1,28 @@
     pass
 
 
+def print_model(model):
```
I'm unsure about the purpose of this. No tests fail if I change `print(model)` to `pass` and print make me nervous that something might go stdout (which doesn't seem to be the case) but at least some comment would help.

AlexHill replied:
Mmm, this is a dummy function to test the "default" error case for a missing lazy operation – that is, one that isn't related to a signal or field. Its body and name are irrelevant – I'll update to make that clearer.

timgraham commented on docs/ref/checks.txt:
```diff
@@ -138,6 +138,9 @@ Models
   together.
+* **models.E022**: ``<function>`` contains a lazy reference to
+  ``<app label>>.<model>``, but app ``<app label>`` is not installed or
+  does not provide ``<model>``.
```
The code says, "doesn't provide" -- let's use "isn't" too".

timgraham commented on docs/ref/checks.txt:
```diff
@@ -237,6 +240,9 @@ Related Fields
 * **fields.W341**: ``ManyToManyField`` does not support ``validators``.
 * **fields.W342**: Setting ``unique=True`` on a ``ForeignKey`` has the same
   effect as using a ``OneToOneField``.
+* **fields.E350**: The field ``<app label>.<model>.<field name>`` was declared
```
what's your thinking about skipping some numbers (342 > 350)?

AlexHill replied:
I noticed looking over the existing codes that they seemed vaguely grouped, with breaks like this introducing new groups of similar checks (E320, E330 etc). Looking over it again it seems this check perhaps belongs with the first group of E300s.

timgraham commented:
About the numbering/categorization questions, I think the main thing to think about it whether or not separate codes are needed for silencing purposes. I guess Django probably isn't very usable if these errors are silenced so it's probably not very important.

AlexHill commented:
Ok, diff since your last comments: http://dpaste.com/2F9Q4QK
* no more `print_model` – I've consolidated the dummy objects in the model validation tests and moved them out of module scope.
* Moved the new field check E350 to E307 to be grouped with similar checks.
* Updated wording in check docs to match code
* Slight tweak to error wording ("The function '<funcname>'" rather than "The '<funcname>' function")
* Rebased onto master and squashed back into two commits

timgraham commented on tests/model_validation/tests.py:
```diff
@@ -29,28 +21,64 @@ def test_models_validate(self):
-                "sender, which has not been installed.",
+                "%s contains a lazy reference to auth.imaginarymodel, "
+                "but app 'auth' doesn't provide model 'imaginarymodel'."
+                % repr(dummy_function),
```
Latest edits look good. Just noticed a small thing: is there any problem using `%r` instead of `%s` and `repr()`?

AlexHill replied:
Didn't know about that! Updated.

AlexHill commented:
I've just added a new test that explicitly fails the new `fields.E307` in `model_validation/tests.py`. The codepath and errors are tested elsewhere (in `migrations/test_state.py`), but I think it makes sense to test all these checks together, in case the migrations tests are changed in the future.

timgraham commented on tests/model_validation/tests.py:
```diff
@@ -29,28 +22,77 @@ def test_models_validate(self):
+        apps.lazy_model_operation(dummy_function, ('auth', 'imaginarymodel'))
+        apps.lazy_model_operation(dummy_function, ('fanciful_app', 'imaginarymodel'))
+
+        post_init.connect(dummy_function, sender='missing-app.Model', apps=apps)
```
Is it important to disconnect these listeners similar to what's done in tests/signals?

AlexHill replied:
`Signal` uses weak references so I think we could get away with it, but it's a good idea. Updated.

AlexHill replied:
Actually these are never even connected, since the sender never exists, which also means we couldn't disconnect them properly anyway. I think it's OK as-is.

That raises a question – should `ModelSignal.disconnect()` use the lazy operation system as well? Currently (both before and after this patch) this happens:

```python
>>> import django
>>> from django.db import models
>>> from django.db.models.signals import post_init
>>> 
>>> django.setup()
>>> 
>>> def receiver(sender, **kwargs):
...     print("post_init sent by %r" % sender)
... 
>>> post_init.connect(receiver, sender='auth.SomeModel')
>>> post_init.disconnect(receiver, sender='auth.SomeModel')
False
>>> 
>>> class SomeModel(models.Model):
...     class Meta:
...         app_label = 'auth'
... 
>>> s = SomeModel()
post_init sent by <class 'SomeModel'>
>>> 
```

I'll open another PR about that.

timgraham commented:
merged in 2ff7ef15b0a1d41e3f121e96cb72a383863046c0 and 779bb82f519420418a711f7ae75e71d600de3224, thanks!
